napari/config.py

```diff
@@ -0,0 +1,877 @@
+            if k not in old or old[k] is None:
+                old[k] = {}
+            update(old[k], v, priority=priority)  # type: ignore
+        else:
+            if priority == "new" or k not in old:
+                old[k] = v
+
+    return old
+
+
+def merge(*dicts: dict) -> dict:
+    """Update a sequence of nested dictionaries.
+
+    This prefers the values in the latter dictionaries to those in the former
+
+    Examples
+    --------
+    >>> a = {'x': 1, 'y': {'a': 2}}
+    >>> b = {'y': {'b': 3}}
+    >>> merge(a, b)
+    {'x': 1, 'y': {'a': 2, 'b': 3}}
+
+    See Also
+    --------
+    napari.config.update
+    """
+    result: dict = {}
+    for d in dicts:
+        update(result, d)
+    return result
+
+
+def collect_yaml(paths: List[str] = paths) -> List[dict]:
+    """Collect configuration from yaml files.
+
+    This searches through a list of paths, expands to find all yaml or json
+    files, and then parses each file.
+    """
+    # Find all paths
+    file_paths = []
+    for path in paths:
+        if os.path.exists(path):
+            if os.path.isdir(path):
+                try:
+                    file_paths.extend(
+                        sorted(
+                            [
+                                os.path.join(path, p)
+                                for p in os.listdir(path)
+                                if os.path.splitext(p)[1].lower()
+                                in (".json", ".yaml", ".yml")
+                            ]
+                        )
+                    )
+                except OSError:
+                    # Ignore permission errors
+                    pass
+            else:
+                file_paths.append(path)
+    configs = []
+
+    # Parse yaml files
+    for path in file_paths:
+        try:
+            with open(path) as f:
+                data = yaml.safe_load(f.read()) or {}
+                configs.append(data)
+        except (OSError, IOError):
+            # Ignore permission errors
+            pass
+
+    return configs
+
+
+def collect_env(env: Optional[Union[dict, os._Environ]] = None) -> dict:
+    """Collect config from environment variables.
+
+    This grabs environment variables of the form "NAPARI_FOO__BAR_BAZ=123" and
+    turns these into config variables of the form ``{"foo": {"bar-baz": 123}}``
+    It transforms the key and value in the following way:
+
+    -  Lower-cases the key text
+    -  Treats ``__`` (double-underscore) as nested access
+    -  Calls ``ast.literal_eval`` on the value
+    """
+    if env is None:
+        env = os.environ
+    d = {}
+    for name, value in env.items():
+        if name.startswith("NAPARI_"):
+            varname = name[7:].lower().replace("__", ".")
+            try:
+                d[varname] = ast.literal_eval(value)
+            except (SyntaxError, ValueError):
+                d[varname] = value
+
+    result: dict = {}
+    set(d, config=result, clean=True)
```

Comment:
Shadowing a built-in is naughty. I suggest using `_set` within this module and aliasing `set = _set` at the bottom, so that other importers can do `config.set`.